Use bulk.SSTWriter to write SSTs on the receiver side

jeffrey-xiao opened the pull request:
The first commit implements the `engine.Writer` interface for `bulk.SSTWriter` and also adapts it to accept an `writeCloseSyncer`. This change allows users to specify if they want to write directly to the disk or to an in-memory buffer.

The second commit creates a new package in `engine` called `sst` that stores `sst.Iterator` and `sst.Writer`.

The third commit uses `bulk.SSTWriter` to write SSTs when receiving snapshots.

Using the same recovery evaluation metric in #38932, I found that there was a nice win in terms of recovery time for an average range size of 128 MiB.

```
name         old s/op   new s/op   delta
AvgBytes121  38.8 ±19%  32.6 ± 7%  -16.03%  (p=0.011 n=8+5)
```

This change should be beneficial across the board since we are avoiding the cost of a Cgo call per `Put`. We are also reducing the memory cost since we are eliminating the in-memory buffer when creating the SST chunks incrementally.

Depends on petermattis/pebble#240.

Fixes #39748.

jeffrey-xiao committed:
bulk: implement writer interface for bulk.SSTWriter

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/39802)
<!-- Reviewable:end -->

jeffrey-xiao committed:
storage: create engine.sst

This commit moves SSTWriter and sstIterator to a new package engine.sst.

Release note: None

jeffrey-xiao committed:
storage: use sst.Writer to write SSTs on the receiver side

This change should be beneficial across the board since we are avoiding
the cost of a Cgo call per Put. We are also reducing the memory cost
since we are eliminating the in-memory buffer when creating the SST
chunks incrementally.

Release note: None

dt reviewed (approved):
I communicated a couple high-level suggestions offline (making a new `sst` package instead of adding more to the `engine` package and a couple naming nits), but general approach LGTM (for the `bulk` parts -- I didn't look at the core parts)

nvb reviewed (commented):
Very nice! Thanks @jeffrey-xiao. It's worth mentioning that another benefit of this change is that you avoid an extra memcpy during the SST writing process because there's no longer an intermediate buffer.

> Reviewed 1 of 3 files at r1, 2 of 2 files at r2, 12 of 12 files at r3, 2 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39802#-:-LnF-X--71XWqQ8Z5CnD:b-t60d6o)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jeffrey-xiao and @nvanbenschoten)

___
*[pkg/storage/store_snapshot.go, line 133 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/39802#-LnEzsCQ91ggRHGPNdqh:-LnEzsCQ91ggRHGPNdqi:bd0dkha) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a5396e9418d903af2ea46d6dc7ef4dd852460a9/pkg/storage/store_snapshot.go#L133)):*
> ```Go
> }
> 
> type sstFile struct {
> ```

Explain what the purpose of this type is. Something about it being an adapter to allow `SSTSnapshotStorageFile` to implement 
___
*[pkg/storage/store_snapshot.go, line 157 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/39802#-LnF0aLS7XnAiz6gRlbt:-LnF0aLS7XnAiz6gRlbu:b-vsb0f0) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a5396e9418d903af2ea46d6dc7ef4dd852460a9/pkg/storage/store_snapshot.go#L157)):*
> ```Go
> }
> 
> func (ssf *sstFile) Sync() error {
> ```

I'm probably missing it, but when is this called? Also, how are we ensuring that the file is synced before closing?
___
*[pkg/storage/store_snapshot.go, line 273 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/39802#-LnF-El3EwsiJa_jeJFE:-LnF-El3EwsiJa_jeJFF:bxojjse) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a5396e9418d903af2ea46d6dc7ef4dd852460a9/pkg/storage/store_snapshot.go#L273)):*
> ```Go
> 	defer func() {
> 		// Closing the same SST multiple times is idempotent.
> 		msstw.Close()
> ```

`defer msstw.Close()`
___
*[pkg/storage/bulk/sst_writer.go, line 157 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39802#-LnEwu7k7lAplZ-eAlTk:-LnEwu7k7lAplZ-eAlTl:b-ybm4i4) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad10afffd8019c3cc7dfec4ba7c030df7698b3ed/pkg/storage/bulk/sst_writer.go#L157)):*
> ```Go
> }
> 
> // ApplyBatchRepr implements the Writer interface.
> ```

Here and below: `engine.Writer interface`, because we're not in the `engine` package.
___
*[pkg/storage/bulk/sst_writer.go, line 189 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39802#-LnExB3r1A7yGnCVmy1g:-LnExB3r1A7yGnCVmy1h:b-fwezkx) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad10afffd8019c3cc7dfec4ba7c030df7698b3ed/pkg/storage/bulk/sst_writer.go#L189)):*
> ```Go
> 	}
> 	fw.DataSize += int64(len(start.Key)) + int64(len(end.Key))
> 	fw.scratch1 = engine.EncodeKeyToBuf(fw.scratch1[:0], start)
> ```

You might as well just use the same scratch buffer instead of introducing a new one:

```
fw.scratch = engine.EncodeKeyToBuf(fw.scratch[:0], start)
startScratch := fw.scratch
fw.scratch = engine.EncodeKeyToBuf(fw.scratch, end)
endScratch := fw.scratch[len(startScratch):]
return fw.fw.DeleteRange(startScratch, endScratch)
```
___
*[pkg/storage/engine/sst/writer.go, line 29 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39802#-LnEyjiV6NLd0LaGVTMO:-LnEyjiV6NLd0LaGVTMP:bt3invr) ([raw file](https://github.com/cockroachdb/cockroach/blob/e22df9f2273e9683df1e1816a8123cd77c5df0f7/pkg/storage/engine/sst/writer.go#L29)):*
> ```Go
> )
> 
> // MakeIntTableKVs returns MVCC key values for a number of keys with given
> ```

Does this need to be part of the package's external interface? What about the next two functions? It looks like they came from a test file, so I don't see a reason why they need to be exported or even included in this file instead of `writer_test.go`.
___
*[pkg/storage/engine/sst/writer.go, line 93 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/39802#-LnF0Ag3CTanZMmaOwOa:-LnF0Ag3CTanZMmaOwOb:b-5e3cwj) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a5396e9418d903af2ea46d6dc7ef4dd852460a9/pkg/storage/engine/sst/writer.go#L93)):*
> ```Go
> }
> 
> type writeCloseSyncer interface {
> ```

I'm a little surprised `govet` isn't complaining that this isn't exported even though it's part of the external interface of the package. Either way, you'll want to export it.


<!-- Sent from Reviewable.io -->

dt reviewed (approved):
> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39802)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jeffrey-xiao and @nvanbenschoten)

___
*[pkg/storage/engine/sst/writer.go, line 29 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39802#-LnEyjiV6NLd0LaGVTMO:-LnF1KCS1VlEWQ2HpED3:b-zc2544) ([raw file](https://github.com/cockroachdb/cockroach/blob/e22df9f2273e9683df1e1816a8123cd77c5df0f7/pkg/storage/engine/sst/writer.go#L29)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Does this need to be part of the package's external interface? What about the next two functions? It looks like they came from a test file, so I don't see a reason why they need to be exported or even included in this file instead of `writer_test.go`.
</blockquote></details>

they're reused `bulk`'s tests as well. I get that _test files aren't linked in non-test, but how I wish go would let _other test binaries_ link them. I think they could get `Testing` added somewhere in their names though.
___
*[pkg/storage/engine/sst/writer.go, line 93 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/39802#-LnF0Ag3CTanZMmaOwOa:-LnF1_cTAvhzMXZSzUAc:b-zesno8) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a5396e9418d903af2ea46d6dc7ef4dd852460a9/pkg/storage/engine/sst/writer.go#L93)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'm a little surprised `govet` isn't complaining that this isn't exported even though it's part of the external interface of the package. Either way, you'll want to export it.
</blockquote></details>

This us just re-stating an un-exported interface that is used in the Pebble public API. I think maybe govet lets you get away with it for interfaces since they're structurally instead of nominatively typed anyway? 


<!-- Sent from Reviewable.io -->

jeffrey-xiao reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39802)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/storage/store_snapshot.go, line 157 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/39802#-LnF0aLS7XnAiz6gRlbt:-LnGI5G--NYIm5f44lbS:b-c35qx1) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a5396e9418d903af2ea46d6dc7ef4dd852460a9/pkg/storage/store_snapshot.go#L157)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'm probably missing it, but when is this called? Also, how are we ensuring that the file is synced before closing?
</blockquote></details>

This is passed into Pebble's `sstable.Writer` and it's called when the sstable is closed. I just realized we would need logic to call `Sync` when a "chunk" has been written. I've added this to `multiSSTWriter`.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 15 of 15 files at r5, 12 of 12 files at r6, 2 of 3 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39802)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/storage/store_snapshot.go, line 157 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/39802#-LnF0aLS7XnAiz6gRlbt:-LnTkJtkBisHm1Xu8QIP:b-j9f4ez) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a5396e9418d903af2ea46d6dc7ef4dd852460a9/pkg/storage/store_snapshot.go#L157)):*
> This is passed into Pebble's sstable.Writer and it's called when the sstable is closed.

Where?

Also, I worry about the performance of calling `limitBulkIOWrite` on every single key write now. Do you think it would make sense to move that to where we call `Sync` instead?
___
*[pkg/storage/bulk/sst_writer.go, line 189 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39802#-LnExB3r1A7yGnCVmy1g:-LnTih5-30oyIW29AQk0:bgi6uo3) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad10afffd8019c3cc7dfec4ba7c030df7698b3ed/pkg/storage/bulk/sst_writer.go#L189)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

You might as well just use the same scratch buffer instead of introducing a new one:

```
fw.scratch = engine.EncodeKeyToBuf(fw.scratch[:0], start)
startScratch := fw.scratch
fw.scratch = engine.EncodeKeyToBuf(fw.scratch, end)
endScratch := fw.scratch[len(startScratch):]
return fw.fw.DeleteRange(startScratch, endScratch)
```
</blockquote></details>

This isn't quite what we want. Since we're passing `fw.scratch[len(startScratch):]` to `EncodeKeyToBuf`, we're going to throw away part of the buffer if it needs to be resized. Instead, we want to append to the entire thing and then slice it up later.
___
*[pkg/storage/engine/sst/writer.go, line 29 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39802#-LnEyjiV6NLd0LaGVTMO:-LnTjnCs8hz2zr-z9vzS:biu08bz) ([raw file](https://github.com/cockroachdb/cockroach/blob/e22df9f2273e9683df1e1816a8123cd77c5df0f7/pkg/storage/engine/sst/writer.go#L29)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

they're reused `bulk`'s tests as well. I get that _test files aren't linked in non-test, but how I wish go would let _other test binaries_ link them. I think they could get `Testing` added somewhere in their names though.
</blockquote></details>

Well at a minimum we should also move these to the bottom of this file. I'd lean towards moving them into their own `writer_testing.go` file.


<!-- Sent from Reviewable.io -->

jeffrey-xiao reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39802)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/storage/store_snapshot.go, line 157 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/39802#-LnF0aLS7XnAiz6gRlbt:-LnVQ_6V2DWcjyd36xkl:b-lh0tet) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a5396e9418d903af2ea46d6dc7ef4dd852460a9/pkg/storage/store_snapshot.go#L157)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> This is passed into Pebble's sstable.Writer and it's called when the sstable is closed.

Where?

Also, I worry about the performance of calling `limitBulkIOWrite` on every single key write now. Do you think it would make sense to move that to where we call `Sync` instead?
</blockquote></details>

The sync is called here: https://github.com/petermattis/pebble/blob/53f7531cb726299eb33b5849ae6935501f357442/sstable/writer.go#L649-L652

Makes sense to me.  `SSTSnapshotFile` would have to keep track of the bytes written before a `Sync` in that case. Does that sound right?
___
*[pkg/storage/bulk/sst_writer.go, line 189 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39802#-LnExB3r1A7yGnCVmy1g:-LnVQzST95w0e9E49rjj:b-n8qqh1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad10afffd8019c3cc7dfec4ba7c030df7698b3ed/pkg/storage/bulk/sst_writer.go#L189)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This isn't quite what we want. Since we're passing `fw.scratch[len(startScratch):]` to `EncodeKeyToBuf`, we're going to throw away part of the buffer if it needs to be resized. Instead, we want to append to the entire thing and then slice it up later.
</blockquote></details>

Hmm, I'm not sure I fully understand. If we pass `fw.scratch` to `engine.EncodeKeyToBuf` when encoding the end key, it's overwriting the buffer instead of append to it. Am I missing something?
___
*[pkg/storage/engine/sst/writer.go, line 29 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39802#-LnEyjiV6NLd0LaGVTMO:-LnVQtBx4KMOWO7_xeAZ:bnpuvbd) ([raw file](https://github.com/cockroachdb/cockroach/blob/e22df9f2273e9683df1e1816a8123cd77c5df0f7/pkg/storage/engine/sst/writer.go#L29)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Well at a minimum we should also move these to the bottom of this file. I'd lean towards moving them into their own `writer_testing.go` file.
</blockquote></details>

Ack. Will move them to `writer_testing.go` when I make the other changes.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
Sorry about letting this sit @jeffrey-xiao. We're ramping up for the next release and it's all hands on deck to stabilize things until then.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39802#-:-Lo6EvTB2f_aqu9b5XyU:b-56vtyf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/storage/store_snapshot.go, line 157 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/39802#-LnF0aLS7XnAiz6gRlbt:-Lo6BA_r2l9vAJTm9lGx:b-zdauug) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a5396e9418d903af2ea46d6dc7ef4dd852460a9/pkg/storage/store_snapshot.go#L157)):*
> The sync is called here: https://github.com/petermattis/pebble/blob/53f7531cb726299eb33b5849ae6935501f357442/sstable/writer.go#L649-L652

Ok, so it's the pebble sst writer itself that calls sync. That's worth a comment right above `if err := fw.fw.Close(); err != nil {`.

> SSTSnapshotFile would have to keep track of the bytes written before a Sync in that case. Does that sound right?

You mean `multiSSTWriter` or `sstFile`? I think it would be most natural to switch the state around to be `bytesSinceLastSync` and then to maintain this on `sstFile`. You would increment it in `sstFile.Write` and clear it in `sstFile.Sync`. This state is a property of a specific file, not the`multiSSTWriter`, so it's strange that it lives there right now.

Does that make sense?
___
*[pkg/storage/bulk/sst_writer.go, line 189 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39802#-LnExB3r1A7yGnCVmy1g:-Lo69F5Q7pKT8gwB0eBl:btc84jr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad10afffd8019c3cc7dfec4ba7c030df7698b3ed/pkg/storage/bulk/sst_writer.go#L189)):*
<details><summary><i>Previously, jeffrey-xiao (Jeffrey Xiao) wrote…</i></summary><blockquote>

Hmm, I'm not sure I fully understand. If we pass `fw.scratch` to `engine.EncodeKeyToBuf` when encoding the end key, it's overwriting the buffer instead of append to it. Am I missing something?
</blockquote></details>

Oh, I see. It's idiomatic for this style of function to append instead of overwrite/re-allocate for exactly this reason. Now we don't know whether the resulting slice is shorter or longer than the original one, so it's not clear whether we want to replace `fw.scratch` or not. It looks like this was already causing confusion because the existing caller to this was passing `fw.scratch[:0]`, which isn't necessary because the length is ignored and only the capacity is consulted.

If you're feeling inspired, I think the right thing to do here would be to have `EncodeKeyToBuf` append the key to the provided buffer instead of doing what it's doing now.
___
*[pkg/storage/engine/sst/writer.go, line 29 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39802#-LnEyjiV6NLd0LaGVTMO:-Lo6AwK72bCliOHAnmf4:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/e22df9f2273e9683df1e1816a8123cd77c5df0f7/pkg/storage/engine/sst/writer.go#L29)):*
<details><summary><i>Previously, jeffrey-xiao (Jeffrey Xiao) wrote…</i></summary><blockquote>

Ack. Will move them to `writer_testing.go` when I make the other changes.
</blockquote></details>

:thumbsup: 


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39802)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/storage/store_snapshot.go, line 157 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/39802#-LnF0aLS7XnAiz6gRlbt:-Lo6FYZu0Lv2vgaL3W1l:b-3sbpcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a5396e9418d903af2ea46d6dc7ef4dd852460a9/pkg/storage/store_snapshot.go#L157)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> The sync is called here: https://github.com/petermattis/pebble/blob/53f7531cb726299eb33b5849ae6935501f357442/sstable/writer.go#L649-L652

Ok, so it's the pebble sst writer itself that calls sync. That's worth a comment right above `if err := fw.fw.Close(); err != nil {`.

> SSTSnapshotFile would have to keep track of the bytes written before a Sync in that case. Does that sound right?

You mean `multiSSTWriter` or `sstFile`? I think it would be most natural to switch the state around to be `bytesSinceLastSync` and then to maintain this on `sstFile`. You would increment it in `sstFile.Write` and clear it in `sstFile.Sync`. This state is a property of a specific file, not the`multiSSTWriter`, so it's strange that it lives there right now.

Does that make sense?
</blockquote></details>

Also take a look at `vfs.NewSyncingFile` which syncs periodically as a file is being written. Writing 64 MB of data and then calling `Sync()` can cause large latency spikes for other concurrent IO operations. We configure RocksDB and Pebble to sync every 512KB.


<!-- Sent from Reviewable.io -->
